### ISSUE TITLE
null to text conversion in array

### ISSUE THREAD

**Issue by AndreyKlimchuk** [epgsql/epgsql#193]
Hello, suppose I have an array of string where some elements are nulls. The result of select of the array is:
```
epgsql:equery(C, "SELECT $1", [[<<"s1">>, null]]).        
{error,{error,error,<<"42P18">>,indeterminate_datatype,
              <<"could not determine data type of parameter $1">>,
              [{file,<<"postgres.c">>},
               {line,<<"1336">>},
               {routine,<<"exec_parse_message">>},
               {severity,<<"ERROR">>}]}}
```
cast fixes above error
```
epgsql:equery(C, "SELECT $1::TEXT[]", [[<<"s1">>, null]]).
{ok,[{column,<<"text">>,{array,text},1009,-1,-1,1}],
    [{[<<"s1">>,<<"null">>]}]}

```
but in the result the null value is also converted to string. How to avoid above error  and preserve type of null values?

**Comment by seriyps** [comment]
That's interesting! I would say, 1st example (without explicit typecast) is like that by design (postgresql client 1st sends only SQL query to postgresql server and asks server to send back the argument types information, but server can't figure out the types from just `SELECT $1`).

But 2nd example is indeed something that can be fixed. Can you, please, try to run your 2nd example on epgsql 3.4.0 to confirm that this issue is a regression from 4.0 major refactoring?

**Comment by seriyps** [comment]
Note for myself

NULL handling in decoder:

https://github.com/epgsql/epgsql/blob/46b353886338c13280eb1ca66c425d9137e38ca2/src/epgsql_wire.erl#L145-L146

NULL handling in encoder:

https://github.com/epgsql/epgsql/blob/46b353886338c13280eb1ca66c425d9137e38ca2/src/epgsql_wire.erl#L247-L250

**Comment by AndreyKlimchuk** [comment]
> That's interesting! I would say, 1st example (without explicit typecast) is like that by design (postgresql client 1st sends only SQL query to postgresql server and asks server to send back the argument types information, but server can't figure out the types from just `SELECT $1`).
> 
> But 2nd example is indeed something that can be fixed. Can you, please, try to run your 2nd example on epgsql 3.4.0 to confirm that this issue is a regression from 4.0 major refactoring?

On 3.4.0 it is not reproducible. Seems like regression.

**Comment by seriyps** [comment]
Ok, thanks! I will try to look at this, I think I have some idea how this should be fixed, but can't guarantee any timeframe when this will be done.
Thank you for your report!

**Comment by AndreyKlimchuk** [comment]
Thank you for the help!

**Comment by seriyps** [comment]
So yes, this was finaly fixed in #212
